hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java

```diff
@@ -304,9 +307,38 @@ private void addResolvedEntityInsertAction(AbstractEntityInsertAction insert) {
 
 	@SuppressWarnings("unchecked")
 	private <T extends Executable & Comparable & Serializable> void addAction(Class<T> executableClass, T action) {
+		if ( action instanceof EntityUpdateAction && trySquashEntityUpdateAction( (EntityUpdateAction) action ) ) {
+			return;
+		}
 		EXECUTABLE_LISTS_MAP.get( executableClass ).getOrInit( this ).add( action );
 	}
 
+	private boolean trySquashEntityUpdateAction(EntityUpdateAction action) {
+		if ( insertions != null ) {
+			for ( NonIdentifierAttribute attr : action.getPersister().getEntityMetamodel().getProperties()) {
+				if ( attr.getType() instanceof EntityType ) {
+					// avoid potential foreign key violation
+					// TODO filter out @ForeignKey(value = ConstraintMode.NO_CONSTRAINT)
```

Comment:
@beikov I fixed foreign key violation here, but it's a little pessimistic to assume all associations have constraints.

Reply:
It seems sessionFactory doesn't retain metadata, there is no way to determine if foreign key actually exists here.